TUnit.Assertions/AssertConditions/AssertionResult.cs

```diff
@@ -0,0 +1,67 @@
+﻿namespace TUnit.Assertions.AssertConditions;
+
+public class AssertionResult
+{
+    public bool IsPassed { get; }
+    public string? Message { get; }
+
+    private AssertionResult(bool isPassed, string? message)
+    {
+        IsPassed = isPassed;
+        Message = message;
+	}
+
+	public static AssertionResult FailIf(Func<bool> isFailed, string message)
+	{
+		if (!isFailed())
+		{
+			return Passed;
+		}
+		return new AssertionResult(false, message);
+	}
+
+	public AssertionResult And(AssertionResult other)
+	{
+		if (IsPassed && other.IsPassed)
+		{
+			return this;
+		}
+
+		if (IsPassed)
+		{
+			return other;
+		}
+
+		if (other.IsPassed)
+		{
+			return this;
+		}
+
+		return Fail(this.Message + " and " + other.Message);
+	}
+
+	public AssertionResult Or(AssertionResult other)
+	{
+		if (!IsPassed && !other.IsPassed)
+		{
+			return Fail(this.Message + " and " + other.Message);
+		}
+
+		return this;
+	}
```

Comment:
I think this is causing your test failures:

```suggestion
	public AssertionResult Or(AssertionResult other)
	{
		if (!IsPassed && !other.IsPassed)
		{
			return Fail(this.Message + " and " + other.Message);
		}

		return Passed;
	}
```